zenith_utils/src/sync.rs

```diff
@@ -0,0 +1,174 @@
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::{io, task};
+
+pin_project! {
+    /// We use this future to mark certain methods
+    /// as callable in both sync and async modes.
+    #[repr(transparent)]
+    pub struct SyncFuture<S, T: Future> {
```

Comment:
The name `SyncFuture` is a bit misleading because it's only sync if `S` is SyncProof. But I couldn't think of a better name, so I can't object.

Reply:
I won't mind if we rename this once we've come up with a better name :)